core/domain/stats_domain.py

```diff
@@ -16,213 +16,256 @@
+            answer_str=submitted_answer_dict.get('answer_str'),
+            large_bucket_entity_id=submitted_answer_dict.get(
+                'large_bucket_entity_id'))
+
+    def validate(self):
+        """Validates this submitted answer object."""
+        # TODO(bhenning): Validate the normalized answer against future answer
+        # objects after #956 is addressed.
+        if self.time_spent_in_sec is None:
+            raise utils.ValidationError(
+                'SubmittedAnswers must have a provided time_spent_in_sec')
+        if self.session_id is None:
+            raise utils.ValidationError(
+                'SubmittedAnswers must have a provided session_id')
+
+        if self.rule_spec_str is not None and not isinstance(
+                self.rule_spec_str, basestring):
+            raise utils.ValidationError(
+                'Expected rule_spec_str to be either None or a string, '
+                'received %s' % str(self.rule_spec_str))
+
+        if self.answer_str is not None and not isinstance(
+                self.answer_str, basestring):
+            raise utils.ValidationError(
+                'Expected answer_str to be either None or a string, received '
+                '%s' % str(self.answer_str))
+
+        if not isinstance(self.session_id, basestring):
+            raise utils.ValidationError(
+                'Expected session_id to be a string, received %s' %
+                str(self.session_id))
+
+        if not isinstance(self.time_spent_in_sec, number.Number):
+            raise utils.ValidationError(
+                'Expected time_spent_in_sec to be a number, received %s' %
+                str(self.time_spent_in_sec))
+
+        if not isinstance(self.params, dict):
+            raise utils.ValidationError(
+                'Expected params to be a dict, received %s' % str(self.params))
+
+        if not isinstance(self.answer_group_index, int):
+            raise utils.ValidationError(
+                'Expected answer_group_index to be an integer, received %s' %
+                str(self.answer_group_index))
+
+        if not isinstance(self.rule_spec_index, int):
+            raise utils.ValidationError(
+                'Expected rule_spec_index to be an integer, received %s' %
+                str(self.rule_spec_index))
+
+        if self.answer_group_index < 0:
+            raise utils.ValidationError(
+                'Expected answer_group_index to be non-negative, received %d' %
+                self.answer_group_index)
+
+        if self.rule_spec_index < 0:
+            raise utils.ValidationError(
+                'Expected rule_spec_index to be non-negative, received %d' %
+                self.rule_spec_index)
+
+        if self.time_spent_in_sec < 0.:
+            raise utils.ValidationError(
+                'Expected time_spent_in_sec to be non-negative, received %f' %
+                self.time_spent_in_sec)
+
+        if self.answer is None and self.interaction_id != 'Continue':
```

Comment:
I was preferring to do this in a follow-up PR after the migration was finished, but did it now.